pkg/kv/kvtest/store.go

```diff
@@ -0,0 +1,269 @@
+	if err != nil {
+		t.Fatalf("failed to set key '%s', to value '%s': %s", testKey, testValue1, err)
+	}
+
+	// get test key with value1
+	val, err := store.Get(ctx, testKey)
+	switch {
+	case err != nil:
+		t.Fatalf("failed to get key '%s': %s", testKey, err)
+	case val == nil:
+		t.Fatalf("got value with nil")
+	case !bytes.Equal(testValue1, val):
+		t.Fatalf("key='%s' value='%s' doesn't match, expected='%s'", testKey, val, testValue1)
+	}
+
+	// override key with value2
+	err = store.Set(ctx, testKey, testValue2)
+	if err != nil {
+		t.Fatalf("failed to set key '%s', to value '%s': %s", testKey, testValue2, err)
+	}
+
+	// get test key with value2
+	val2, err := store.Get(ctx, testKey)
+	if err != nil {
+		t.Fatalf("failed to get key '%s': %s", testKey, err)
+	}
+	if !bytes.Equal(testValue2, val2) {
+		t.Fatalf("key='%s' value='%s' doesn't match, expected='%s'", testKey, val2, testValue2)
+	}
+
+	// get a missing key
+	const keyNotExists = "key-not-exists"
+	val3, err := store.Get(ctx, []byte(keyNotExists))
+	if !errors.Is(err, kv.ErrNotFound) {
+		t.Fatalf("get key='%s' err=%s, expected not found", keyNotExists, err)
+	}
+	if val3 != nil {
+		t.Fatalf("get key='%s' value='%s', expected nil", keyNotExists, val3)
+	}
+}
+
+func testStoreDelete(t *testing.T, ms makeStore) {
+	ctx := context.Background()
+	store := ms(t, ctx)
+	defer store.Close()
+
+	t.Run("exists", func(t *testing.T) {
+		keyToDel := []byte("key-to-delete")
+		err := store.Set(ctx, keyToDel, []byte("value to delete"))
+		if err != nil {
+			t.Fatalf("failed to set key='%s': %s", keyToDel, err)
+		}
+		err = store.Delete(ctx, keyToDel)
+		if err != nil {
+			t.Fatalf("failed to delete key='%s': %s", keyToDel, err)
+		}
+	})
+
+	t.Run("non_exists", func(t *testing.T) {
+		keyToDel := []byte("missing-key-to-delete")
+		err := store.Delete(ctx, keyToDel)
+		if !errors.Is(err, kv.ErrNotFound) {
+			t.Fatalf("delete missing key '%s', err=%v expected=%s", keyToDel, err, kv.ErrNotFound)
+		}
+	})
+}
+
+func testStoreSetIf(t *testing.T, ms makeStore) {
+	ctx := context.Background()
+	store := ms(t, ctx)
+	defer store.Close()
+
+	t.Run("previous_value", func(t *testing.T) {
+		// setup initial value
+		key := []byte("setIfKey1")
+		value := []byte("val1")
+		err := store.Set(ctx, key, value)
+		if err != nil {
+			t.Fatalf("failed to set key=%s, value=%s: %s", key, value, err)
+		}
+
+		// should set new value if old value is kept
+		newValue := []byte("val2")
+		err = store.SetIf(ctx, key, newValue, value)
+		if err != nil {
+			t.Fatalf("failed to setIf key=%s value=%s pred=%s: %s", key, newValue, value, err)
+		}
+
+		// should NOT set new value if old value is not kept
+		newValue = []byte("val3")
+		err = store.SetIf(ctx, key, newValue, value)
+		if !errors.Is(err, kv.ErrNotFound) {
+			t.Fatalf("setIf err=%s, key=%s, value=%s, pred=%s, expected err=%s", err, key, newValue, value, kv.ErrNotFound)
+		}
+
+		// should NOT set new value if try conditional nil
```

Comment:
What _is_ the desired behaviour for a `nil` conditional?  If it is always to fail then I would rather make it an application error.

Reply:
The case if setting a key first. Passing nil as precondition is to make sure there is currently no value set.

Reply:
Cool.  Then let's document it!

I think it would be easier for me to read if this appeared in the next test (which tests behaviour with old value `nil`); perhaps rename it to `OldValueNil`?  That way each case in the function documentation is tested by one function.  Your call, though.